anomaly-benchmarking/generate-datasets/script.whizzml

```diff
@@ -0,0 +1,205 @@
+        ano-var ((fetch ano-cl-id) ["clusters"
+                                    "global"
+                                    "distance"
+                                    "standard_deviation"])
+        my-ratio (pow (/ norm-var ano-var) 2)
+        new-tag (cond (< my-ratio 0.25) "high scatter"
+                      (< my-ratio 0.5) "medium scatter"
+                      (< my-ratio 1) "low scatter"
+                      (< my-ratio 2) "low clusteredness"
+                      (< my-ratio 4) "medium clusteredness"
+                      "high clusteredness"))
+    (when delete-resources
+          (map safe-delete [ano-cl-id]))
+    (update ano-ds-id {"tags" (append my-tags new-tag)})))
+
+;; Takes a dataset. Determines which relative frequencies are
+;; possible. Generates the max possible datasets (limit 40) at those
+;; frequencies. Returns a list of datasets.
+
+(define (relative-frequency ds-id normal-rows)
+  (let (ds (fetch ds-id)
+        availible-rows (ds "rows")
+        ds-name (ds "name")
+        my-tags (ds "tags")
+        max-freq (/ availible-rows (+ availible-rows normal-rows)))
+    (for (freq [0.001 0.005 0.01 0.05 0.1])
+      (let (k (/ (* freq normal-rows) (- 1 freq)))
+        (when (> max-freq freq)
+              (repeatedly (min 10 (floor (/ availible-rows k))) 
+                          (lambda () (create-dataset {"origin_dataset" ds-id
+                                       "sample_rate" (/ k availible-rows)
+                                       "name" (str ds-name " - " freq)
+                                       "tags" (append my-tags (str freq))}))))))))
+  
+;; Given an anomalous dataset and a normal dataset, combines them into
+;; a single dataset, and returns a map with key/value pairs from the
+;; tags from the anomalous dataset (difficulty, frequency, and
+;; variation) and the combined dataset (resource)
+
+(define (generate-map ds-id normal-ds)
+  (let (ds (fetch ds-id)
+        combined (create-dataset {"origin_datasets" [ds-id normal-ds]}) 
+        my-tags (ds "tags"))
+    (make-map ["difficulty" "frequency" "variation"  "resource"] 
+              (append my-tags combined))))
+
+;; Putting it all together. This function takes a dataset, makes the
+;; dataset binary using the make-binary library, and splits it into
+;; parts by point-difficulty, creates a high and low clusteredness for
+;; each part, and generates as many frequencies as possible for each
+;; clusteredness. Returns a list of maps for all the datasets created,
+;; specifying the resource id, difficulty, semantic variation, and
+;; frequency.
+
+(define (generate ds-id delete-resources)
+  (let (old-obj ((fetch ds-id) ["objective_field" "id"])
+        binary (make-binary ds-id old-obj delete-resources)
+        new-obj ((fetch binary) ["objective_field" "id"])
+        normal-ds (create-and-wait-dataset {"origin_dataset" binary
+                    "lisp_filter" 
```

Comment:
Sorry about all the indentation errors. It's a combination of me getting used to formatting conventions, and wanting more space as my expressions get longer and longer. I won't let it happen again. At least it's not tabs? :)